src/GeoModel.js

```diff
@@ -1,95 +1,290 @@
 import BaseModel from './BaseModel'
 import Promise from 'bluebird'
 
-import traverse from 'traverse'
 import _ from "underscore"
+import rewind from 'geojson-rewind'
+import turf from '@turf/meta'
+import turfCentroid from '@turf/centroid'
+turf.centroid = turfCentroid
 
+/** Cleans up polygon coordinates.
+ * For each linear ring that defines the polygon, it removes repeated
+ * coordinates, ensures ring is closed. It discards rings with less than
+ * four points, and if the outer ring is discarded the polygon becomes
+ * undefined.
+ * @private
+ */
 function cleanPolygon(coordinates) {
-  return _.uniq(coordinates, false, (i) => { return i[0]+"-"+i[1] }).concat([coordinates[0]])
-  
+  function near(v1, v2) { return Math.abs(v1 - v2) < 0.000001 } // about 10 cm
+  function same(c1, c2) { return near(c1[0], c2[0]) && near(c1[1], c2[1]) }
+
+  let rslt = []
+  coordinates.forEach((ring, index) => {
+    ring = ring.reduce((acc, val, idx) => {
+      if ((idx === 0) || !same(acc[acc.length - 1], val)) acc.push(val)
+      return acc
+    }, [])
+    if (same(ring[0], ring[ring.length - 1])) ring.pop()
+    ring.push(ring[0])
+    if (ring.length < 4) ring = undefined
+    if (ring || (index === 0)) rslt.push(ring)
+  })
+  return rslt[0] && rslt
 }
 
-function extractGeometry(obj) {
-  return traverse(obj).reduce(function(acc, x) {
-    if(this.key == 'geometry') {
-      if (x.type == 'Polygon') {
-        // Filter illegal repeats of coordinates, re-add end point
-        x.coordinates[0] = cleanPolygon(x.coordinates[0])
-      }
-      if (x.type == 'MultiPolygon') {
-        // Filter illegal repeats of coordinates, re-add end point
-        x.coordinates = _.map(x.coordinates, (coords) => {
-          coords[0] = cleanPolygon(coords[0])
-          return coords
-        })
-      }
-      acc.push(x)
-    };
-    return acc;
-  }, [])
+function cleanGeometryCoordinates(parts) {
+  let cleaned = []
+  parts.Polygon.forEach((coordinates) => {
+    coordinates = cleanPolygon(coordinates)
+    if (coordinates) {
+      let obj = rewind({type: 'Polygon', coordinates})
+      cleaned.push(obj.coordinates)
+    }
+  })
+  parts.Polygon = cleaned
 }
 
-function extractGeometryFeatures(values, next) {
-  var orig = this.geometryField
-  var features = this.geometryFeatureField
-  if (values[orig]) {
-    let val = values[orig]
-    if (_.isString(val)) {
-      val = JSON.parse(val)
+/** Extracts geometry coordinates from a GeoJSON object.
+ * It is strict in that it processes only portions of the object that it
+ * recognizes as valid GeoJSON, but lenient in that it accomodates
+ * missing properties and ignores misplaced or unrecognized properties.
+ * @private
```

Comment:
I think you could be right – some of these _might_ be useful as static methods. However, my bias is to leave them private until we have a demonstrated need.